### PR TITLE
Drawing datastructure

### PR DESCRIPTION
**Pull request by mvanmeerten**

### CODE REVIEW

File: src/main/java/programminglife/gui/controller/GraphController.java [hunk 1]
```diff
@@ -36,117 +35,79 @@ public GraphController(GenomeGraph graph, Group grpDrawArea) {
...
-                        .setY(INITIAL_OFFSET.getY() + (int) (CHILD_OFFSET * childCount * node.getHeight()));
-                this.drawDFS(node, child, newOffset, maxDepth - 1, drawnNodes);
-                childCount++;
+        for (DrawableNode drawableNode : subGraph.getNodes().values()) {
+            drawNode(drawableNode);
+            for (DrawableNode child : subGraph.getChildren(drawableNode)) {
+                drawEdge(drawableNode, child);
             }
         }
+        long finishTime = System.nanoTime();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Do we want these benchmarks in the release?

File: src/main/java/programminglife/gui/controller/GraphController.java [hunk 2]
```diff
@@ -36,117 +35,79 @@ public GraphController(GenomeGraph graph, Group grpDrawArea) {
...
+
+        edge.setOnMouseClicked(event -> {
+            Console.println(edge.toString());
+        });
+
+        edge.setStroke(Color.DARKGRAY);
+        edge.setStrokeWidth(3);
+        edge.setStartLocation(edge.getStart().getRightBorderCenter());
+        edge.setEndLocation(edge.getEnd().getLeftBorderCenter());
+        Console.println(edge.toString());
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/gui/controller/GraphController.java]: This was for debug and should be removed.

File: src/main/java/programminglife/gui/controller/GraphController.java [hunk 3]
```diff
@@ -36,117 +35,79 @@ public GraphController(GenomeGraph graph, Group grpDrawArea) {
...
-        this.grpDrawArea.getChildren().add(new Rectangle(1, 2, 3, 4));
-
-        return node.getSize();
+        drawableNode.setFill(Color.TRANSPARENT);
+        drawableNode.setStroke(Color.BLUE);
+        if (drawableNode.getNode() instanceof Dummy) {
+            drawableNode.setStroke(Color.DARKGRAY);
+            drawableNode.setStrokeWidth(3);
+        }
+       // System.out.println(drawableNode.toString());
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Not needed anymore, can be removed.

File: src/main/java/programminglife/model/Dummy.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public int getIdentifier() {
+        return id;
+    }
+
+    @Override
+    public Collection<? extends Edge> getChildEdges() {
+        Collection result = new HashSet<Link>();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Dummy.java]: I believe a LinkedHashSet is faster when you only have a small amount of entries.

File: src/main/java/programminglife/model/Dummy.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        result.add(new Link(this, child, child.getGenomes()));
+        return result;
+    }
+
+    @Override
+    public Collection<? extends Edge> getParentEdges() {
+        Collection result = new HashSet<Link>();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Dummy.java]: I believe a LinkedHashSet is faster when you only have a small amount of entries.

File: src/main/java/programminglife/model/Dummy.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        result.add(new Link(parent, this, this.getGenomes()));
+        return result;
+    }
+
+    @Override
+    public Collection<? extends Node> getChildren() {
+        Collection result = new HashSet();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Dummy.java]: I believe a LinkedHashSet is faster when you only have a small amount of entries.

File: src/main/java/programminglife/model/Dummy.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        result.add(child);
+        return result;
+    }
+
+    @Override
+    public Collection<? extends Node> getParents() {
+        Collection result = new HashSet();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Dummy.java]: I believe a LinkedHashSet is faster when you only have a small amount of entries.

File: src/main/java/programminglife/model/GenomeGraph.java [hunk 8]
```diff
@@ -33,7 +33,7 @@ public GenomeGraph(String id) {
 
     /**
      * The constructor for a GenomeGraph.
-     * @param id String.
+     * @param id String.úú
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/GenomeGraph.java]: typo

File: src/main/java/programminglife/gui/controller/GraphController.java [hunk 9]
```diff
@@ -36,117 +35,79 @@ public GraphController(GenomeGraph graph, Group grpDrawArea) {
...
+        drawableNode.setOnMouseClicked(event -> {
+            Console.println(drawableNode.getSequence());
+            Console.println(drawableNode.toString());
         });
 
-        node.setFill(Color.TRANSPARENT);
-        node.setStroke(Color.DARKRED);
-
-        this.grpDrawArea.getChildren().add(node);
-        this.grpDrawArea.getChildren().add(new Rectangle(1, 2, 3, 4));
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Do we want this to get our graph bit more back to the middle.

File: src/main/java/programminglife/model/Segment.java [hunk 10]
```diff
@@ -92,98 +79,65 @@ public int getIdentifier() {
...
-     */
     @Override
-    public String toString() {
-        String sequence = this.getSequence();
-        return String.format("Segment<%d>[s(%d):%s]",
-                this.getIdentifier(),
-                sequence.length(),
-                StringUtils.abbreviate(sequence, 11));
+    public Collection<? extends Edge> getChildEdges() {
+        Collection<Link> result = new HashSet<>();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Segment.java]: I believe a LinkedHashSet is faster when you only have a small amount of entries.

File: src/main/java/programminglife/model/Segment.java [hunk 11]
```diff
@@ -92,98 +79,65 @@ public int getIdentifier() {
...
-    /**
-     * Get a {@link XYCoordinate} representing the size of the {@link Segment}.
-     * @return The size of the {@link Segment}
-     */
-    public XYCoordinate getSize() {
-        if (!drawDimensionsUpToDate) {
-            setDrawDimensions();
+    @Override
+    public Collection<? extends Edge> getParentEdges() {
+        Collection<Link> result = new HashSet<>();
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/Segment.java]: I believe a LinkedHashSet is faster when you only have a small amount of entries.

File: src/main/java/programminglife/gui/controller/GraphController.java (hunk 1 above)
Comment (mvanmeerten) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Could be useful to know.

File: src/main/java/programminglife/gui/controller/GraphController.java (hunk 2 above)
Comment (mvanmeerten) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Removed in next commit. Is now on-click.

File: src/main/java/programminglife/gui/controller/GraphController.java (hunk 9 above)
Comment (mvanmeerten) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Feels a bit dirty to do that.

File: src/main/java/programminglife/model/drawing/DrawableNode.java [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        }
+        return this.getCenter().add(this.getSize().getX() / 2, 0);
+    }
+
+    /**
+     * Method to return a string with information about the {@link DrawableNode}.
+     * @return a {@link String} representation of a {@link DrawableNode}.
+     */
+    @Override
+    public String toString() {
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/drawing/DrawableNode.java]: Maybe we want a better toString. Depends if we use this with the onClick.

File: src/main/java/programminglife/gui/controller/GraphController.java (hunk 3 above)
Comment (mvanmeerten) [inline src/main/java/programminglife/gui/controller/GraphController.java]: This is what to draw for a dummy node. It is necessary.

File: src/main/java/programminglife/model/drawing/SubGraph.java [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+     * This will put each node in a Layer one higher than each of its parents.
+     * @return A {@link List} of Layers with all the nodes (all nodes are divided over the Layers).
+     */
+    private List<Layer> findLayers() {
+        long startTime = System.nanoTime();
+        List<DrawableNode> sorted = topoSort();
+        long finishTime = System.nanoTime();
+        long differenceTime = finishTime - startTime;
+        long milisecondTime = differenceTime / 1000000;
+        Console.println("TIMERINO OF TOPOSORTERINO:  " + milisecondTime);
```
Comment (IwanHoogenboom) [inline src/main/java/programminglife/model/drawing/SubGraph.java]: This is debug things.

File: src/main/java/programminglife/model/Dummy.java (hunk 4 above)
Comment (mvanmeerten) [inline src/main/java/programminglife/model/Dummy.java]: Not currently used, so we don't know what is faster.

File: src/main/java/programminglife/model/Dummy.java (hunk 5 above)
Comment (mvanmeerten) [inline src/main/java/programminglife/model/Dummy.java]: Not currently used, so we don't know what is faster.

File: src/main/java/programminglife/gui/controller/GraphController.java (hunk 9 above)
Comment (IwanHoogenboom) [inline src/main/java/programminglife/gui/controller/GraphController.java]: Yeah a bit

**Pull request merged**